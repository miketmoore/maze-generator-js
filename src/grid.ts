import { ICoord, coordFactory } from './coord'
import { ICell, cellFactory } from './cell'
import { Direction } from './direction'
import { randInRange } from './rand'
import { IWall } from './wall'

type ExportCell = [string, boolean, boolean, boolean, boolean]

export interface ExportData {
  readonly grid: ExportCell[]
}

export interface IGrid {
  readonly getCell: (coord: ICoord) => ICell | undefined
  readonly getAdjacentCoord: (
    direction: Direction,
    coord: ICoord
  ) => ICoord | undefined
  readonly getRandCoord: () => ICoord
  readonly getAvailableCellWalls: (cell: ICell, cellCoord: ICoord) => IWall[]
  readonly getExportData: () => ExportData
}

class Grid implements IGrid {
  private rows: number
  private cols: number
  private cells: Record<string, ICell>
  constructor(rows: number, cols: number) {
    this.rows = rows
    this.cols = cols
    this.cells = {}
    for (let row = 0; row < rows; row++) {
      for (let col = 0; col < cols; col++) {
        const key = this.buildKey(row, col)
        this.cells[key] = cellFactory()
      }
    }
  }

  public getCell = (coord: ICoord) => {
    const { row, col } = coord
    const key = this.buildKey(row, col)
    return this.cells[key]
  }

  private buildKey = (row: number, col: number) => `${row},${col}`

  private getAdjacentCellCoords: (
    direction: Direction,
    coord: ICoord
  ) => ICoord | undefined = (direction, coord) => {
    let new_row = coord.row
    let new_col = coord.col
    switch (direction) {
      case 'north':
        if (coord.row === 0) {
          return undefined
        }
        new_row--
        break
      case 'east':
        if (coord.col === this.cols - 1) {
          return undefined
        }
        new_col++
        break
      case 'south':
        if (coord.row === this.rows - 1) {
          return undefined
        }
        new_row++
        break
      case 'west':
        if (coord.col === 0) {
          return undefined
        }
        new_col--
        break
    }
    return coordFactory(new_row, new_col)
  }

  public getAdjacentCoord: (
    direction: Direction,
    coord: ICoord
  ) => ICoord | undefined = (direction, coord) => {
    const adjacentCoords = this.getAdjacentCellCoords(direction, coord)
    if (!adjacentCoords) {
      return undefined
    }
    return this.coordInBounds(adjacentCoords) ? adjacentCoords : undefined
  }

  private rowInBounds = (row: number) => row >= 0 && row < this.rows
  private colInBounds = (col: number) => col >= 0 && col < this.cols
  private coordInBounds = (coord: ICoord) => {
    return this.rowInBounds(coord.row) && this.colInBounds(coord.col)
  }

  public getRandCoord = () =>
    coordFactory(randInRange(0, this.rows - 1), randInRange(0, this.cols - 1))

  private isWallAvailable = (cellCoord: ICoord, wall: IWall) => {
    if (wall.isSolid()) {
      const adjacentCoord = this.getAdjacentCoord(
        wall.getDirection(),
        cellCoord
      )
      if (adjacentCoord) {
        const adjacentCell = this.getCell(adjacentCoord)
        if (adjacentCell && !adjacentCell.isVisited()) {
          return true
        }
      }
    }
    return false
  }
  // available cell walls are walls that have not been carved and that are adjacent to a cell
  // that has not been visited
  public getAvailableCellWalls = (cell: ICell, cellCoord: ICoord) =>
    cell
      .getWalls()
      .toArray()
      .filter(wall => this.isWallAvailable(cellCoord, wall))

  public getExportData = () => {
    const data: ExportData = { grid: [] }
    Object.keys(this.cells).forEach(key => {
      const cell = this.cells[key]
      const walls = cell.getWalls()
      data.grid.push([
        key,
        walls.north.isSolid(),
        walls.east.isSolid(),
        walls.south.isSolid(),
        walls.west.isSolid()
      ])
    })
    return data
  }
}

export const gridFactory: (rows: number, cols: number) => IGrid = (
  rows,
  cols
) => new Grid(rows, cols)
