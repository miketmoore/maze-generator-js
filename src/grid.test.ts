import { gridFactory, IGrid } from './grid'
import { coordFactory } from './coord'

let grid: IGrid
beforeEach(() => (grid = gridFactory(4, 4)))

describe('grid', () => {
  describe('getCell', () => {
    const validCoords = [[0, 0], [1, 2], [3, 3]]
    validCoords.forEach(([row, col]) => {
      test(`returns a cell at coordinate [row:${row}, col:${col}]`, () => {
        const coord = coordFactory(row, col)
        const cell = grid.getCell(coord)
        expect(cell).toBeDefined()
      })
    })

    const invalidCoords = [[-1, -1], [4, 4]]
    invalidCoords.forEach(([row, col]) => {
      test(`returns undefined for invalid coordinates`, () => {
        const coord = coordFactory(row, col)
        const cell = grid.getCell(coord)
        expect(cell).toBeUndefined()
      })
    })
  })
  describe('getAdjacentCoord', () => {
    test('returns an adjacent cell', () => {
      const coord = coordFactory(0, 0)
      const adjacent = grid.getAdjacentCoord('east', coord)
      if (!adjacent) {
        throw new Error('expected adjacent coord to be defined')
      }
      expect(adjacent.row).toEqual(0)
      expect(adjacent.col).toEqual(1)
    })
    test('does not return a cell', () => {
      const coord = coordFactory(3, 3)
      const adjacent = grid.getAdjacentCoord('east', coord)
      expect(adjacent).toBeUndefined()
    })
  })
  describe('getExportData', () => {
    test('returns an un-carved grid', () => {
      expect(grid.getExportData()).toMatchSnapshot()
    })
    test('returns an un-carved grid', () => {
      const carveData: Array<
        [number, number, Array<'north' | 'east' | 'south' | 'west'>]
      > = [
        [0, 0, ['north']],
        [1, 2, ['north', 'east']],
        [2, 3, ['north', 'west']],
        [3, 3, ['south']]
      ]
      carveData.forEach(([row, col, wallNames]) => {
        const cell = grid.getCell(coordFactory(row, col))
        if (!cell) {
          throw new Error('cell is undefined which is unexpected')
        }
        const walls = cell.getWalls()
        wallNames.forEach(wallName => walls[wallName].carve())
      })

      expect(grid.getExportData()).toMatchSnapshot()
    })
  })
})
