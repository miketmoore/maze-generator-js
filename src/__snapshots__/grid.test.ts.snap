// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`grid getExportData returns an un-carved grid 1`] = `
Object {
  "grid": Array [
    Array [
      "0,0",
      true,
      true,
      true,
      true,
    ],
    Array [
      "0,1",
      true,
      true,
      true,
      true,
    ],
    Array [
      "0,2",
      true,
      true,
      true,
      true,
    ],
    Array [
      "0,3",
      true,
      true,
      true,
      true,
    ],
    Array [
      "1,0",
      true,
      true,
      true,
      true,
    ],
    Array [
      "1,1",
      true,
      true,
      true,
      true,
    ],
    Array [
      "1,2",
      true,
      true,
      true,
      true,
    ],
    Array [
      "1,3",
      true,
      true,
      true,
      true,
    ],
    Array [
      "2,0",
      true,
      true,
      true,
      true,
    ],
    Array [
      "2,1",
      true,
      true,
      true,
      true,
    ],
    Array [
      "2,2",
      true,
      true,
      true,
      true,
    ],
    Array [
      "2,3",
      true,
      true,
      true,
      true,
    ],
    Array [
      "3,0",
      true,
      true,
      true,
      true,
    ],
    Array [
      "3,1",
      true,
      true,
      true,
      true,
    ],
    Array [
      "3,2",
      true,
      true,
      true,
      true,
    ],
    Array [
      "3,3",
      true,
      true,
      true,
      true,
    ],
  ],
}
`;

exports[`grid getExportData returns an un-carved grid 2`] = `
Object {
  "grid": Array [
    Array [
      "0,0",
      false,
      true,
      true,
      true,
    ],
    Array [
      "0,1",
      true,
      true,
      true,
      true,
    ],
    Array [
      "0,2",
      true,
      true,
      true,
      true,
    ],
    Array [
      "0,3",
      true,
      true,
      true,
      true,
    ],
    Array [
      "1,0",
      true,
      true,
      true,
      true,
    ],
    Array [
      "1,1",
      true,
      true,
      true,
      true,
    ],
    Array [
      "1,2",
      false,
      false,
      true,
      true,
    ],
    Array [
      "1,3",
      true,
      true,
      true,
      true,
    ],
    Array [
      "2,0",
      true,
      true,
      true,
      true,
    ],
    Array [
      "2,1",
      true,
      true,
      true,
      true,
    ],
    Array [
      "2,2",
      true,
      true,
      true,
      true,
    ],
    Array [
      "2,3",
      false,
      true,
      true,
      false,
    ],
    Array [
      "3,0",
      true,
      true,
      true,
      true,
    ],
    Array [
      "3,1",
      true,
      true,
      true,
      true,
    ],
    Array [
      "3,2",
      true,
      true,
      true,
      true,
    ],
    Array [
      "3,3",
      true,
      true,
      false,
      true,
    ],
  ],
}
`;
